### PR TITLE
[NCCL] Timeout Loop Thread for Async Error Handling

### PR DESCRIPTION
**Pull request by osalpekar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44163 [NCCL] Add Environment Variable to guard Async Error Handling feature
* #41054 [NCCL] Destructor Blocks on WorkNCCL Completion
* #41053 [NCCL] WorkNCCL Helper Functions
* #41052 [NCCL] Abort Errored and Timed Out NCCL Communicators from Watchdog Thread
* #41051 [NCCL] Use cudaEventQuery to Poll for GPU operation errors
* **#41050 [NCCL] Timeout Loop Thread for Async Error Handling**

**This Commit:**
We introduce a workVector to track live workNCCL objects corresponding to collective operations. Further, we introduce a workCleanupLoop, which busy-polls the vector of workNCCL objects and removes them upon completion.

**This Stack:**
The purpose of this stack is to fix the hanging behavior observed in when using PyTorch DDP training with NCCL. In various situations (desynchronization, high GPU utilization, etc.), NCCL collectives may hang due to waiting on an unresponsive worker. This stack detects such hanging behavior and aborts timed-out collectives by throwing a user-visible exception, all with minimal perf regression. Training can then be restarted from a previous checkpoint with something like torchelastic.

Differential Revision: [D21916637](https://our.internmc.facebook.com/intern/diff/D21916637/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21916637/)!

### CODE REVIEW

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -778,6 +807,16 @@ c10::intrusive_ptr<c10::ivalue::Future> ProcessGroupNCCL::WorkNCCL::
   return c10::make_intrusive<FutureNCCL>(shared_from_this(), outputs_);
 }
 
+void ProcessGroupNCCL::enqueue(
+    std::shared_ptr<ProcessGroupNCCL::WorkNCCL> work) {
+  {
+    std::lock_guard<std::mutex> lock(workVectorMutex_);
+    auto it = workVector_.end();
+    workVector_.emplace(it, std::move(work));
```
Comment (mrshenli) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:815]: curious, is this faster than `workVector_.emplace(std::move(work))`?

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -778,6 +807,16 @@ c10::intrusive_ptr<c10::ivalue::Future> ProcessGroupNCCL::WorkNCCL::
   return c10::make_intrusive<FutureNCCL>(shared_from_this(), outputs_);
 }
 
+void ProcessGroupNCCL::enqueue(
```
Comment (mrshenli) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:810]: (can be add in a followup PR if necessary) I wonder if this enqueue/workCleanupLoop/runLoop can be abstract out and deduped  from ProcessGroupGloo/NCCL and put it into the base class?

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 3]
```diff
@@ -542,7 +546,32 @@ void ProcessGroupNCCL::ncclCommWatchdogInternal() {
...
+
+void ProcessGroupNCCL::workCleanupLoop() {
+  while (!terminateProcessGroup_.load()) {
+    std::unique_lock<std::mutex> lock(workVectorMutex_);
+    // We busy-poll the work vector every kWatchdogThreadSleepMillis
+    // milliseconds as long as the atomic is True.
+    workVectorCV_.wait_for(
+        lock,
+        std::chrono::milliseconds(kWatchdogThreadSleepMillis),
+        [&]() -> bool { return terminateProcessGroup_.load(); });
```
Comment (mrshenli) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:561]: Is this correct? This means this won't wake up until `terminateProcessGroup_` is true, even if there are notifications?

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 2 above)
Comment (osalpekar) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:810]: There is some NCCL-specific logic added to `workCleanupLoop` on in the stack, but we have some ideas on potentially consolidating the looping threads to reduce the overhead of additional threads if we have multiple process groups. I believe the priority for this work item depends on the results from the multiple process groups experiments for TSC.

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (osalpekar) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:815]: I think `std::vector::emplace` requires a position argument, but we can use `std::vector::emplace_back` here instead.

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 3 above)
Comment (osalpekar) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:563]: Yeah I think we didn't notify the `workVectorCV` from enqueue in an earlier version - seems like it's best to remove it entirely and always sleep for the specified duration instead

**Review by pritamdamania87 — APPROVED** [review]
There seem to be some DDP tests failing, please check those before landing.

**Pull request closed without merging**